### ISSUE TITLE
[ECS] [request]: Task Definition Life Cycle Management

### ISSUE THREAD

**Issue by corddyla** [aws/containers-roadmap#1967]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Tell us about your request**

<span textcolor="#24292f" style="--original-color: #24292f;color: #24292f">***Request for comment: Please share your thoughts on this proposed improvement to Amazon ECS.***</span><br><br>ECS will deliver a smarter/automated mechanism for task definition lifecycle management. Customers will be able to configure task definition revision lifecycle policies at an account level to automatically deregister and delete task definition revisions that are no longer in use. With this feature, you would be able to configure how long to retain task definition revisions with an active or inactive state that are not associated with any running task or service before they are either deregistered or deleted. If you have not opted into configuring a retention period based on time in days since last used for an active task definition revision, you can configure how many active past revisions per task definition family you would like to retain before those task definition revisions are deregistered.<br><br><br>Proposed API spec for the lifecycle policy is:<br><br><pre>{<br>   "PutTaskDefinitionLifeCyclePolicy": { <br>      "ActiveTaskDefinitionRetentionPeriod": integer,<br>      "InactiveTaskDefinitionRetentionPeriod": integer,<br>      "MaximumRevisionsPerTaskDefinitionFamily": integer<br>   }<br>}</pre>

Parameter | Definition
-- | --
ActiveTaskDefinitionRetentionPeriod | This  attribute is defined in terms of days and sets the maximum for how many days a task  definition revision with an active state that are not associated with any  running task or service should be retained before they are deregistered. This attribute is compared to the days since the task definition revision was last used by a service or task. This  attribute can only be configured if  MaximumRevisionsPerTaskDefinitionFamily  is not configured.This  attribute is specified in terms of days and can be specified as a value from  0 through 365 (where zero means forever).
InactiveTaskDefinitionRetentionPeriod | This  attribute is defined in terms of days and sets the maximum for how long task  definition revisions with an inactive state (already deregistered) that are  not associated with any running task or service should be retained before  they are deleted. This attribute is compared to the days since the task definition revision was deregistered. This  attribute is specified in terms of days and can be specified as a value from  0 through 365 (where zero means forever).
MaximumRevisionsPerTaskDefinitionFamily | This  attribute sets the maximum number of active task definition revisions that should  be retained before they are deregistered. The oldest revisions based on last used date will be deregistered. This  attribute can only be configured if  ActiveTaskDefinitionRetentionPeriod  is not configured.  This  attribute is specified in terms of the number of task definition revisions  and can be specified as a value from 0 through N (where zero means unlimited).

<br>For example:<br><br>Case 1:<br><ul><li>If you used a value of 180 for <code>ActiveTaskDefinitionRetentionPeriod</code>, ECS will deregister any active task definition revision that are not associated with any running tasks or services and have not been used for 180 days.</li><li>If you used a value of 180 for <br><code>InactiveTaskDefinitionRetentionPeriod</code>, ECS will delete any inactive task definition revision that are not associated with any running tasks or services and have been deregistered more than 180 days ago</li></ul>Case 2: <br><code>ActiveTaskDefinitionRetentionPeriod</code> and <br><code>MaximumRevisionsPerTaskDefinitionFamily</code> are mutually exclusive and thus you can only configure one at any given point<br><ul><li>If you used a value of 10 for <code>MaximumRevisionsPerTaskDefinitionFamily</code>, ECS will only retain the 10 latest active revisions in a task definition family and will deregister any others that are not associated with any running task or service.</li></ul>

**Which service(s) is this request for?**
ECS

**Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?**
I need a scalable way to manage my task definition resources.

**Are you currently working around this issue?**
Task Definitions can be managed (registered, deregistered, and deleted) using ECS Console or via API.

**Additional context**
https://github.com/aws/containers-roadmap/issues/685

**Attachments**
N/A

**Comment by JohnPreston** [comment]
The only issue I see here, is for the rollback feature, when enabled, the retention needs to be ignored though. The rollback, iirc, is set on the service :thinking: 
If for a reason a service has to rollback to previous definitions, does that mean that whilst the setting for rollback is on, none of these should be automatically cleaned up?

**Comment by corddyla** [comment]
> The only issue I see here, is for the rollback feature, when enabled, the retention needs to be ignored though. The rollback, iirc, is set on the service 🤔 If for a reason a service has to rollback to previous definitions, does that mean that whilst the setting for rollback is on, none of these should be automatically cleaned up?

Hey John, for rollbacks do you only have a need to rollback to the last version? Given that, would ensuring we retain at least two revisions solve for that use-case?

**Comment by rorylogue** [comment]
Has there been any movement on this on any other threads?

We have been using serverless to deploy our ECS task definitions and they're stacking up. When listing task definitions, the AWS API doesn't actually list the full amount because there's too many.

When we specify `var arn = tasks.TaskDefinitionArns.Last();` it's not actually pulling the last now, it pulls the last of the amount it can pull in one call.

I can put in extra logic to call these in batches and then select the highest but that's increasing runtime and just making things over-complicated.

**Comment by davidkshepherd** [comment]
Hi @corddyla, It has been a minute.  
Has there been any movement on this request?  
We are deregistering our aged TaskDefinitions, but it is imperfect at best.

**Comment by rwlodarczyk-xealth** [comment]
We have tons (read: thousands) of old ECS task definitions; would love to establish a lifecycle policy to only keep those created in the last N, say 90, days for purposes of rollback if needed. Timeline for this feature request would be much appreciated.

**Comment by dancorne** [comment]
Yes please! Without an automatic way to remove task definitions it's easy to accrue thousands.

Removing them is made worse by the fact that the DeregisterTaskDefinition API only allows 1 at a time, leading to heavy ECS rate limiting on the account.
